cli/src/cli_util.rs

```diff
@@ -2707,6 +2708,14 @@ fn resolve_aliases(
             real_commands.insert(alias.to_string());
         }
     }
+    for alias in aliases_map.keys() {
+        if real_commands.contains(alias) {
```

Comment:
Alternatively, this check could be moved inside the substitution loop where type error or recursion is detected. I'm not sure which is better, though.

Reply:
the difference is whether the checking is eager or lazy. the substitution map only handles arguments that were passed to the current invocation of jj, and the logic is already quite complicated, i would prefer not to modify it.

(i need to change it for https://github.com/martinvonz/jj/issues/2866 and it's *hairy*)

Reply:
> the difference is whether the checking is eager or lazy.

Yes, that's the point, and I'm not sure which is better. Apparently we don't check the value type eagerly.

Reply:
That said, for `revset`/`template-aliases`, parse error in declaration part (key) is checked eagerly, and error in definition (value) is lazily. So it's somewhat consistent that command names are checked eagerly.

Reply:
@jyn514 what do you think about making it a warning? One advantage of making it a warning is that it lets users define an alias that won't break when the user upgrades to a version of jj that has a command with the same name.

Reply:
i've changed it to give a warning.

> One advantage of making it a warning is that it lets users define an alias that won't break when the user upgrades to a version of jj that has a command with the same name.

i don't think this is true? this keeps the same behavior from before, that aliases cannot override built-ins. so if the user upgrades to a jj that has a new subcommand, their command will change behavior. this just makes it so that they have some indication of what happened.

Reply:
It's been a while and I don't remember what I meant. A warning would have let users define a `next` alias before we had the `next` command, but that's only useful for very simple commands like that.